### ISSUE TITLE
Is there support for Google OAuth 2.0 ?

### ISSUE THREAD

**Issue by gabriel-tentaculo** [react-native-webview/react-native-webview#1761]
Hi there, recently we received an email from Google at work saying that one of the apps we develop may be blocked from signing in because Google is making an effort to block sign in from inside embedded browsers.

![image](https://user-images.githubusercontent.com/70529565/100388146-9275d700-3008-11eb-854b-2ab88e817dcc.png)

Our app uses react-native-webview to render a web page that has html form to login using email and password or social login with Google and others

Here is the link that came in the email
https://developers.googleblog.com/2020/08/guidance-for-our-effort-to-block-less-secure-browser-and-apps.html

And here is the key point from that link
```
If you're an app developer and use CEF or other clients for authorization on devices, 
use browser-based OAuth 2.0 flows. 
Alternatively, you can use a compatible full native browser for sign-in.

For limited-input device applications, such as applications that do not have access to 
a browser or have limited input capabilities, use limited-input device OAuth 2.0 flows.
```
I'm gathering information about this and I wonder if you guys can help me understand if react-native-webview qualifies as one of the cases that will be blocked by Google in their security efforts

Here are two more links that may be helpful understanding Google's pre requisites
https://developers.google.com/identity/protocols/oauth2/native-app
https://security.googleblog.com/2018/10/announcing-some-security-treats-to.html

**Comment by CaptainJeff** [comment]
I'm interested in the same question. Did you figure anything out @gabriel-tentaculo ?

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff Not yet, first of all it's really hard to insert the custom headers test as the blog post below recommends:
https://developers.googleblog.com/2020/08/guidance-for-our-effort-to-block-less-secure-browser-and-apps.html

If you search the issues of this webview project, you will see many people talking about how you can only insert a custom header on the first request, the login system that my app uses it's actually managed first by cognito, with google oauth as a plugin, so it's even harder to trace things as there is a bunch of back and forth communication throughout the login process.

What i'm doing right now is trying what the blog post recommends on chrome browser, but it seems like there is nothing coming from google response indicating that your test (by adding the custom header) was positive or negative, I have just spent a couple of hours on chrome adding and removing theses headers for nothing... 
Bbtw if you wanna go by that route (validating by doing the headers test), I recommend creating your own http server and monitor first if you can successfully add the headers by inspecting the traffict, this article is great on helping you to do that: https://blog.bigbinary.com/2016/07/26/passing-request-headers-on-each-webview-request-in-react-native.html

I have contacted @jamonholmgren about this and he cannot confirm or deny if the recpective webviews of each os (that is what's actually being used by this package) is actually meeting Google standards for OAuth2 or not, and no one I have talked to so far can confirm or deny. I'm trying to figure this out for 3 days now with no avail so far.

Let me know if you want to collaborate on this so we can either find a positive or negative feedback, I'm worried as my client can be badly impacted by this...

**Comment by CaptainJeff** [comment]
@gabriel-tentaculo Yeah would love to collaborate on this and get it knocked out.

I am having the same difficulty in finding out if my practices comply with the google requirements. 

I'm currently considering using one of these two projects

https://www.npmjs.com/package/react-native-inappbrowser-reborn

Or 

https://github.com/FormidableLabs/react-native-app-auth

The inappbrowser from the looks of it seems like its the same issue with webview but it says "
```
Provides access to the system's web browser and supports handling redirects
Chrome Custom Tabs for Android & SafariServices/AuthenticationServices for iOS. 
```

Which seems sufficient to get this working. It would be much easier to use `react-native-inappbrowser-reborn` than `react-native-app-auth` what are your thoughts?

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff glad to hear you're up for some colab!

I'm gonna review your alternatives soon and get back to you.

I just did a quick experiment though that I found to be interesting, maybe relevant to our case...

try setting the URI of the webview to https://www.youtube.com and add the headers and see whats gonna be loaded versus without the headers... When I added the header, I could not acess youtube at all, when I removed it, the youtube page loaded but with a warning box about browser not supported.

Another thing that may be helpful, although I did not even started to look int it is this Google OAuth Playground tool:
https://developers.google.com/oauthplayground/

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff Just an update about this youtube test, unfortunately I cannot reproduce it anymore, maybe I'd need to reset my AVD to test it again, do you mind testing this and see if you find something interesting?

In the meanwhile I plan on creating a fresh account on google apis and sign up for OAth and create a small node server to help me figure this out...

**Comment by gabriel-tentaculo** [comment]
Also, something I just realized when reading the guidance [blog post from the link on the first post](https://developers.googleblog.com/2020/08/guidance-for-our-effort-to-block-less-secure-browser-and-apps.html) for the thousandth time, is that you actually can use these embedded frameworks and browser clients, it says there clearly, all you have to make sure is that you use the OAuth2.0 flow:
![image](https://user-images.githubusercontent.com/70529565/101644497-02765b00-3a14-11eb-85b0-2fa6effd6088.png)


Would that not be the nail in the coffin for this discussion?
Any thoughts @CaptainJeff ?
@jamonholmgren do you have any experience or can you recall any post/issue here where people successfully implemented OAuth2.0 flow as per these docs form Google? https://developers.google.com/identity/protocols/oauth2

**Comment by CaptainJeff** [comment]
@gabriel-tentaculo sorry for my late response! I keep on reading the google blogs and understand them differently every-time. The screenshot you posted does seem to suggest its fine as long as the authentication process follows the 0Auth 2.0 flow but its strange that they don't mention that in the other areas. But I agree, that does seem like as long as the authentication follow these steps https://developers.google.com/identity/protocols/oauth2 we should be okay! 

I'm going to confirm with the rest of my team that they agree with this assessment

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff yea I feel the same as you, every time you read about the subject it seems like my understanding of it changes. 

I started a thread on reddit about the subject too, and I got some interesting replies if you wanna take a look to consider in your assessment: https://www.reddit.com/r/androiddev/comments/k9rh2m/google_oauth_to_block_embedded_browsers_does_it/

**Comment by CaptainJeff** [comment]
@gabriel-tentaculo yeah those are some interesting thoughts and returns us back the cycle of more information makes this harder to understand. 

My team agreed with our assessment that we feel good about using the embedded browser as long as the webview handles the oauth2.0 protocol. Going this route means we can't test it ahead of time though because the headers don't matter. Its a gamble but its the one we're gonna go with! Have you made any decision yet?

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff We are going to follow the same path, a bit of a gamble I agree but after all those days researching we fell confident to go for it, specially since those announcements ("Google will block webview for Signin") exist since 2017/2018, and yet here we are in 2020 using the webview for it with no problems.

One thing that I can recommend you to do is requesting the extension for this security enforcement, which if Google approves it, somehow they will postpone this to June 30th, you can follow this link for that:
https://docs.google.com/forms/d/e/1FAIpQLSdtYRVJSTKUtnqGcE2L9SaY6xgAgy-eHMqqcqIWOqB084Ao0A/viewform

We got this link on the email that Google sent us (check the last paragraph on the print on my first post here) because we are already using this system so this must have triggered some of their tools to send us this warning, in your case, as I understand, you are still to implement this, so the sooner you create your account and start using it in production, I guess the better so you can hit production and ask for this extension right away, or if you are feeling lucky, just go for it already :)

Best of luck for us, and don't hesitate on messaging me if you wanna check up on how we are doing after Jan 5th or any other help needed.

**Comment by CaptainJeff** [comment]
@gabriel-tentaculo WOW! Great suggestion on that extension request! Thanks so much. Lets get in touch after this Jan 4th deadline and we'll be able to make a more informed decision for the future

**Comment by gabriel-tentaculo** [comment]
Sure thing :+1: 
Thanks for helping out :)

**Comment by gabriel-tentaculo** [comment]
hey @CaptainJeff , just hopping back here to see how that worked out for you. 
We have continued using the webview just as before and asked for that time extension, which was granted, and so far so good.
I'm really guessing that by implementing the auth flow as they recommended in their OAuth2 docs have made it be ok in the end.
Hope you had similar luck.

**Comment by CaptainJeff** [comment]
@gabriel-tentaculo 

Thanks for following up! I've been meaning to reach out to you. It has continued to work for us as well and we didn't request the time extension. So it appears as though implementing the oauth2.0 protocol was sufficient!

**Comment by gabriel-tentaculo** [comment]
@CaptainJeff Yes I'd also say that this is the case indeed. 
Now I can sleep knowing that this finally got settled :)
Thanks for replying too.
Best,
